### PR TITLE
core(lightwallet): remove tolerance from timing budgets API

### PR DESCRIPTION
**Pull request by khempenius**
**Summary**
This updates the existing, un-implemented API for timing budgets.

```
// Summary of Timing Budget API proposed in this PR

Budget.TimingBudget {
  metric: TimingMetric;
  budget: number;
}

TimingMetric = 'first-contentful-paint' | 'first-cpu-idle' | 'interactive' | 'first-meaningful-paint' | 'max-potential-fid' | 'estimated-input-latency' | 'total-blocking-time';
```
**Comments**

- I think that `measurementStrategy` and `runs` is a better way to deal with variance than setting a `tolerance`, hence its removal.

**Related Issues**
(same issue, filed two different places)
- https://github.com/GoogleChrome/lighthouse/issues/8917
- https://github.com/GoogleChrome/budget.json/issues/3

**Related PRs**
I will update this with the PR that implements this.

### CODE REVIEW

**Comment by patrickhulce** [comment]
Thanks @khempenius!

> measurementStrategy?: MeasurementStrategy; 


Also, in LHCI we currently call this `mergeMethod`, I'm not necessarily tied to that name but `measurementStrategy` feels a little more like it's trying to describe how to measure each metric as opposed to how to combine multiple measurements.

WDYT of any of these alternatives? (and potentially moving to 'min'/'max' in place of 'optimistic'/'pessimistic' if we replace with statistical*)

- `statisticalMethod`
- `statisticalStrategy` 
- `statisticalApproach`
- `combinationMethod`
- `combinationStrategy` 

> I think that measurementStrategy and runs is a better way to deal with variance than setting a tolerance, hence its removal.

Strongly agree with dropping `tolerance` 👍 SGTM!

> Adding runs to budget.json complicates interactions with LighthouseCI a bit: if someone sets multiple runs while running Lighthouse CI (e.g. —numberOfRuns=5) as well as in budget.json (e.g. runs: 3),

I think `runs` is a field that is only ever going to apply in LHCI anyhow (i.e. Lighthouse core won't be running multiple times), so IMO we just make LHCI override it's default `numberOfRuns` with the budget.json `runs` if it is set.

**Comment by patrickhulce** [comment]
Oh for `measurementStrategy` it's also worth noting that for CDS we plan on introducing 'median-run' which bases the assertion on the value of the `median-run` according to TTI rather than the median of all observed values. This is the method that will be consistent with the LHCI UI.

**Comment by khempenius** [comment]
`statisticalApproach` sounds good to me. I will switch to that.

I  think I prefer `statistical*` over `combination` or `merge` because I think the latter could be misinterpreted as relating to budget/property merging once `budget.json` supports `extends`.

**Comment by khempenius** [comment]
Interesting. Would `median` and `median-run` be two separate strategies then? Or would only `median-run` be supported?

**Comment by patrickhulce** [comment]
> Would median and median-run be two separate strategies then? Or would only median-run be supported?

Two separate strategies. `median` is useful on it's own but for the UI we're only going to be comparing the median run reports so we want something that reflects that strategy as well if you want to be consistent between the two.

**Comment by patrickhulce** [comment]
> statisticalApproach sounds good to me. I will switch to that.
I think I prefer statistical* over combination or merge because I think the latter could be misinterpreted as relating to budget/property merging once budget.json supports extends.

Makes sense! Should we add anything more verbose to ensure they're connected to the statistics of multiple run measurements? Nothing jumps out that isn't super super long to me though `statisticalRunCombinationApproach`, etc haha

Also thoughts on moving to `min`/`max` then if we're reframing as a statistical approach instead of a looser "merge method"? I'm somewhat split between them. I think `min`/`max` requires more thinking on which one you should be using but is will never require explanation of how it works compared to `optimistic`/`pessimistic`.

**Comment by khempenius** [comment]
I think I'm probably very slightly opposed to moving to `median/min/max` because it locks in the assumption that all performance metrics are good if low; bad if high. OTOH, there’s little precedent for a formal metric working opposite to this; a LH score is probably the only counterexample I can think of.

**Comment by patrickhulce** [comment]
> I think I'm probably very slightly opposed to moving to median/min/max because it locks in the assumption that all performance metrics are good if low; bad if high. OTOH, there’s little precedent for a formal metric working opposite to this; a LH score is probably the only counterexample I can think of.

Ah, for budgets it's done only at the file level rather than individual metric/assertion level I suppose, hmmmm. This is a tricky one then 🤔

**Comment by khempenius** [comment]
I really really like `aggregation*`. I think it is a much closer approximation than "combination". `aggregationMethod` is probably my favorite then.

**Comment by patrickhulce** [comment]
Sounds good I'll align on `aggregationMethod` as well! :D

File: types/budget.d.ts [hunk 1]
```diff
@@ -38,12 +38,19 @@ declare global {
...
+       * 'Median' mode fails if the median run fails.
+       * 'Optimistic' mode fails only if all runs fail. Put another way, this passes if any run passes.
+       * 'Pessimistic' mode fails if any of the runs fail. Put another way, this only passes if all runs pass.
+       *  This can be supplied if only 1 run is used, but it is superfluous because all strategies would yield the same result.
+       */
+      export type AggregationMethod = 'median' | 'optimistic' | 'pessimistic';
+
       /** Supported timing metrics. */
-      export type TimingMetric = 'first-contentful-paint' | 'first-cpu-idle' | 'interactive' | 'first-meaningful-paint' | 'max-potential-fid';
+      export type TimingMetric = 'first-contentful-paint' | 'first-cpu-idle' | 'interactive' | 'first-meaningful-paint' | 'max-potential-fid' | 'estimated-input-latency' | 'total-blocking-time';
```
Comment (patrickhulce) [inline types/budget.d.ts]: any intention for speed-index to be captured here? :)

File: types/budget.d.ts [hunk 2]
```diff
@@ -38,12 +38,19 @@ declare global {
...
       }
 
+      /**
+       * This indicates how measurements taken over multiple runs should be interpretted; the default is 1.
+       * 'Median' mode fails if the median run fails.
+       * 'Optimistic' mode fails only if all runs fail. Put another way, this passes if any run passes.
+       * 'Pessimistic' mode fails if any of the runs fail. Put another way, this only passes if all runs pass.
+       *  This can be supplied if only 1 run is used, but it is superfluous because all strategies would yield the same result.
+       */
+      export type AggregationMethod = 'median' | 'optimistic' | 'pessimistic';
```
Comment (patrickhulce) [inline types/budget.d.ts]: seems like we don't actually leverage this at the moment, is it missing from `Budget` interface or did we land on postponing this?

Comment (khempenius) [reply types/budget.d.ts]: Oops, I thought I had removed this when I was removing the `runs` related stuff. Yep, it's not needed.

**Comment by paulirish** [comment]
> we should probably call this out as a breaking change somewhere in the 6.0 docs (not sure if anyone was using this yet, though)

afaik it was undocumented

i'm cool with not calling it breaking.

btw i love this approach. 👍

**Pull request merged**